text/0000-pbft-consensus.md

```diff
@@ -0,0 +1,480 @@
+   messages are legitimate if:
+  + `signer_id` and `summary` of block inside `pre_prepare` match the
+    corresponding fields of the original `BlockNew` block &&
+  + View in `pre_prepare` message corresponds to this server's current view &&
+  + This message hasn't been accepted already with a different `summary` &&
+  + Sequence number is within the sequential bounds of the log (low and high
+    water marks)
+
+   Once the `pre_prepare` is accepted, broadcast a `prepare` message to all nodes.
+
+3. Receive `prepare` messages, and check them all against their associated
+   `pre_prepare` message in this node's message log.
+
+4. Once the predicate `prepared` is true for this node, then call
+   `check_block`. If an error occurs (`ReceiveError` or `UnknownBlock`),
+   abort (call `cancel_block`). Otherwise, wait for a response (`BlockValid`
+   or `BlockInvalid`) from the validator. If `BlockValid`, then broadcast a
+   `commit` message to all other nodes. If `BlockInvalid`, then call
+   `cancel_block`.
+
+5. When the predicate `committed` is true for this node, then send a
+   `commit_final` message back to the primary.
+
+6. Once the primary has accumulated `f + 1` `commit_final` messages, then
+   it should commit the block using `commit_block`, and advance the chain
+   head.
+
+7. If *block duration* has elapsed, then try to `summarize_block` with the
+   current working block. If the working block is not ready, (`BlockNotReady`
+   or `InvalidState` occurs), then nothing happens (call `ignore_block`).
+   Otherwise, `finalize_block` is called, and a new working block is created by
+   calling `initialize_block`.
+
+Messages passed during normal operation are roughly described by the following
+diagram:
+
+![PBFT operation](../images/pbft_sawtooth.png)
+
+One additional message type that may be encountered during the event loop is
+`BlockCommit`. This message type will cause the consensus algorithm to abort its
+current working block, and initialize a new one.
+
+### View Changes
+[view-changes]: #view-changes
+Sometimes, the node currently in charge (the primary) becomes faulty. In this
+case, a view change is necessary. View changes are triggered by a timeout:
+When a secondary node receives a `BlockNew` message, a timer is started. If
+the secondary ends up receiving a `commit_final` message, the timer is
+cancelled, and the algorithm proceeds as normal. If the timer expires, the
+primary node is considered faulty and a view change is initiated. This ensures
+Byzantine fault tolerance due to the fact that each step of the algorithm will
+not proceed to the next unless it receives a certain number of matching
+messages, and due to the fact that the Validator does not pass on any messages
+that have invalid signatures [[3]](#references).
+
+The view change process is as follows:
+1. Any node who discovers the primary as faulty (whose timer timed out) sends
+   a `view_change` message to all nodes.
+
```

Comment:
Should we specify the data that needs to be included in a view_change method, which I believe is current node state (seq. no?) and pending requests.

Reply:
The `view_change` messages include the sequence number, current view, proof of previous checkpoint, and pending requests - this is described under the "Data Structures" section, but I agree it would be good to remind readers here.